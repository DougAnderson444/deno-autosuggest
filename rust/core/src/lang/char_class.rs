use std::fmt;
use super::Lang;


pub trait CharPattern: fmt::Debug {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool>;
}


#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum CharClass {
    Any,
    Control,
    Whitespace,
    Punctuation,
    NotAlphaNum,
    Consonant,
    Vowel,
}

use CharClass::{
    Any,
    Control,
    Whitespace,
    Punctuation,
    NotAlphaNum,
    Consonant,
    Vowel,
};


impl CharPattern for CharClass {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> {
        match self {
            Any         => Some(true),
            Control     => Some(ch.is_control()),
            Whitespace  => Some(ch.is_whitespace()),
            Punctuation => Some(ch.is_ascii_punctuation()),
            NotAlphaNum => Some(!ch.is_alphanumeric()),
            Consonant   => Some(lang.as_ref()?.get_char_class(ch)? == Consonant),
            Vowel       => Some(lang.as_ref()?.get_char_class(ch)? == Vowel),
        }
    }
}


impl<P: CharPattern> CharPattern for [P] {
    fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> {
        let mut met_none = false;
        for pattern in self {
            match pattern.matches(ch, lang) {
                Some(true)  => return Some(true),
                Some(false) => continue,
                None => {
                    met_none = true;
                    continue;
                },
            }
        }
        if met_none { None } else { Some(false) }
    }
}


impl<P: CharPattern> CharPattern for [P; 1] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 2] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 3] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 4] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }
impl<P: CharPattern> CharPattern for [P; 5] { fn matches(&self, ch: char, lang: &Option<Lang>) -> Option<bool> { self[..].matches(ch, lang) } }


#[cfg(test)]
mod tests {
    use super::{CharClass, CharPattern, Lang};

    use CharClass::{
        Any,
        Control,
        Whitespace,
        Punctuation,
        NotAlphaNum,
        Consonant,
        Vowel,
    };

    #[test]
    fn pattern_matches_any() {
        assert_eq!(Any.matches('\0', &None), Some(true));
        assert_eq!(Any.matches('2',  &None), Some(true));
        assert_eq!(Any.matches('f',  &None), Some(true));
        assert_eq!(Any.matches(';',  &None), Some(true));
        assert_eq!(Any.matches(' ',  &None), Some(true));
    }

    #[test]
    fn pattern_matches_control() {
        assert_eq!(Control.matches('\0', &None), Some(true));
        assert_eq!(Control.matches('2',  &None), Some(false));
        assert_eq!(Control.matches('f',  &None), Some(false));
        assert_eq!(Control.matches(';',  &None), Some(false));
        assert_eq!(Control.matches(' ',  &None), Some(false));
    }

    #[test]
    fn pattern_matches_whitespaces() {
        assert_eq!(Whitespace.matches(' ',  &None), Some(true));
        assert_eq!(Whitespace.matches('\t', &None), Some(true));
        assert_eq!(Whitespace.matches('\n', &None), Some(true));
        assert_eq!(Whitespace.matches('2',  &None), Some(false));
        assert_eq!(Whitespace.matches('f',  &None), Some(false));
        assert_eq!(Whitespace.matches(';',  &None), Some(false));
        assert_eq!(Whitespace.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_punctuation() {
        assert_eq!(Punctuation.matches(';',  &None), Some(true));
        assert_eq!(Punctuation.matches('.',  &None), Some(true));
        assert_eq!(Punctuation.matches(',',  &None), Some(true));
        assert_eq!(Punctuation.matches('2',  &None), Some(false));
        assert_eq!(Punctuation.matches('f',  &None), Some(false));
        assert_eq!(Punctuation.matches(' ',  &None), Some(false));
        assert_eq!(Punctuation.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_non_alphanum() {
        assert_eq!(NotAlphaNum.matches('2',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('f',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('ы',  &None), Some(false));
        assert_eq!(NotAlphaNum.matches('も', &None), Some(false));
        assert_eq!(NotAlphaNum.matches(';',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches('.',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches(',',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches(' ',  &None), Some(true));
        assert_eq!(NotAlphaNum.matches('\0', &None), Some(true));
    }

    #[test]
    fn pattern_matches_slice() {
        let pattern = &[Whitespace, Punctuation][..];

        assert_eq!(pattern.matches(' ',  &None), Some(true));
        assert_eq!(pattern.matches(';',  &None), Some(true));
        assert_eq!(pattern.matches('f',  &None), Some(false));
        assert_eq!(pattern.matches('2',  &None), Some(false));
        assert_eq!(pattern.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_array_2() {
        let pattern = [Whitespace, Punctuation];

        assert_eq!(pattern.matches(' ',  &None), Some(true));
        assert_eq!(pattern.matches(';',  &None), Some(true));
        assert_eq!(pattern.matches('f',  &None), Some(false));
        assert_eq!(pattern.matches('2',  &None), Some(false));
        assert_eq!(pattern.matches('\0', &None), Some(false));
    }

    #[test]
    fn pattern_matches_vowel_no_lang() {
        assert_eq!(Vowel.matches('2',  &None), None);
        assert_eq!(Vowel.matches('f',  &None), None);
        assert_eq!(Vowel.matches('ы',  &None), None);
        assert_eq!(Vowel.matches('も', &None), None);
        assert_eq!(Vowel.matches(';',  &None), None);
        assert_eq!(Vowel.matches('.',  &None), None);
        assert_eq!(Vowel.matches(',',  &None), None);
        assert_eq!(Vowel.matches(' ',  &None), None);
        assert_eq!(Vowel.matches('\0', &None), None);
    }

    #[test]
    fn pattern_matches_vowel_with_lang() {
        let mut lang = Lang::new();
        lang.add_char_class('f', Consonant);
        lang.add_char_class('ы', Vowel);
        let lang = Some(lang);

        assert_eq!(Vowel.matches('2',  &lang), None);
        assert_eq!(Vowel.matches('f',  &lang), Some(false));
        assert_eq!(Vowel.matches('ы',  &lang), Some(true));
        assert_eq!(Vowel.matches('も', &lang), None);
        assert_eq!(Vowel.matches(';',  &lang), None);
        assert_eq!(Vowel.matches('.',  &lang), None);
        assert_eq!(Vowel.matches(',',  &lang), None);
        assert_eq!(Vowel.matches(' ',  &lang), None);
        assert_eq!(Vowel.matches('\0', &lang), None);
    }

    #[test]
    fn pattern_matches_consonant_no_lang() {
        assert_eq!(Consonant.matches('2',  &None), None);
        assert_eq!(Consonant.matches('f',  &None), None);
        assert_eq!(Consonant.matches('ы',  &None), None);
        assert_eq!(Consonant.matches('も', &None), None);
        assert_eq!(Consonant.matches(';',  &None), None);
        assert_eq!(Consonant.matches('.',  &None), None);
        assert_eq!(Consonant.matches(',',  &None), None);
        assert_eq!(Consonant.matches(' ',  &None), None);
        assert_eq!(Consonant.matches('\0', &None), None);
    }

    #[test]
    fn pattern_matches_consonant_with_lang() {
        let mut lang = Lang::new();
        lang.add_char_class('f', Consonant);
        lang.add_char_class('ы', Vowel);
        let lang = Some(lang);

        assert_eq!(Consonant.matches('2',  &lang), None);
        assert_eq!(Consonant.matches('f',  &lang), Some(true));
        assert_eq!(Consonant.matches('ы',  &lang), Some(false));
        assert_eq!(Consonant.matches('も', &lang), None);
        assert_eq!(Consonant.matches(';',  &lang), None);
        assert_eq!(Consonant.matches('.',  &lang), None);
        assert_eq!(Consonant.matches(',',  &lang), None);
        assert_eq!(Consonant.matches(' ',  &lang), None);
        assert_eq!(Consonant.matches('\0', &lang), None);
    }
}
