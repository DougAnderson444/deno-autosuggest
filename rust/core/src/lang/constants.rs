use super::CharClass;
use CharClass::{Consonant, Vowel};

pub const CHAR_CLASSES_LATIN: &[(CharClass, char)] = &[
    (Consonant, 'b'),
    (Consonant, 'c'),
    (Consonant, 'd'),
    (Consonant, 'f'),
    (Consonant, 'g'),
    (Consonant, 'h'),
    (Consonant, 'j'),
    (Consonant, 'k'),
    (Consonant, 'l'),
    (Consonant, 'm'),
    (Consonant, 'n'),
    (Consonant, 'p'),
    (Consonant, 'q'),
    (Consonant, 'r'),
    (Consonant, 's'),
    (Consonant, 't'),
    (Consonant, 'v'),
    (Consonant, 'w'),
    (Consonant, 'x'),
    (Consonant, 'z'),
    (Vowel, 'a'),
    (Vowel, 'e'),
    (Vowel, 'i'),
    (Vowel, 'o'),
    (Vowel, 'u'),
    (Vowel, 'y'),
];
