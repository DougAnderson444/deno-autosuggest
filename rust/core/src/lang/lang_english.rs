#![allow(dead_code)]

use rust_stemmers::{Algorithm, Stemmer};
use super::PartOfSpeech;
use super::Lang;

use PartOfSpeech::{
    Article,
    Preposition,
    Conjunction,
    Particle,
};


const FUNCTION_WORDS: &[(PartOfSpeech, &'static str)] = &[
    // Articles
    (Article, "a"),
    (Article, "an"),
    (Article, "the"),

    (Preposition, "at"),
    (Preposition, "by"),
    (Preposition, "for"),
    (Preposition, "from"),
    (Preposition, "in"),
    (Preposition, "of"),
    (Preposition, "on"),
    (Preposition, "to"),
    (Preposition, "since"),
    (Preposition, "before"),
    (Preposition, "till"),
    (Preposition, "untill"),
    (Preposition, "beside"),
    (Preposition, "under"),
    (Preposition, "below"),
    (Preposition, "over"),
    (Preposition, "above"),
    (Preposition, "across"),
    (Preposition, "through"),
    (Preposition, "into"),
    (Preposition, "towards"),
    (Preposition, "onto"),
    (Preposition, "off"),
    (Preposition, "out"),
    (Preposition, "about"),

    (Conjunction, "after"),
    (Conjunction, "although"),
    (Conjunction, "as"),
    (Conjunction, "because"),
    (Conjunction, "before"),
    (Conjunction, "but"),
    (Conjunction, "either"),
    (Conjunction, "for"),
    (Conjunction, "how"),
    (Conjunction, "if"),
    (Conjunction, "lest"),
    (Conjunction, "nor"),
    (Conjunction, "once"),
    (Conjunction, "once"),
    (Conjunction, "or"),
    (Conjunction, "since"),
    (Conjunction, "since"),
    (Conjunction, "since"),
    (Conjunction, "so"),
    (Conjunction, "than"),
    (Conjunction, "that"),
    (Conjunction, "that"),
    (Conjunction, "though"),
    (Conjunction, "till"),
    (Conjunction, "unless"),
    (Conjunction, "until"),
    (Conjunction, "until"),
    (Conjunction, "until"),
    (Conjunction, "what"),
    (Conjunction, "whatever"),
    (Conjunction, "when"),
    (Conjunction, "when"),
    (Conjunction, "whenever"),
    (Conjunction, "where"),
    (Conjunction, "whereas"),
    (Conjunction, "whereas"),
    (Conjunction, "wherever"),
    (Conjunction, "whether"),
    (Conjunction, "which"),
    (Conjunction, "whichever"),
    (Conjunction, "while"),
    (Conjunction, "while"),
    (Conjunction, "while"),
    (Conjunction, "whilst"),
    (Conjunction, "who"),
    (Conjunction, "whoever"),
    (Conjunction, "whom"),
    (Conjunction, "whomever"),
    (Conjunction, "whose"),
    (Conjunction, "why"),
    (Conjunction, "yet"),
    (Conjunction, "and"),
    // (Conjunction, "as if"),
    // (Conjunction, "as long as"),
    // (Conjunction, "as much as"),
    // (Conjunction, "as soon as"),
    // (Conjunction, "as though"),
    // (Conjunction, "assuming that"),
    // (Conjunction, "by the time"),
    // (Conjunction, "even if"),
    // (Conjunction, "even though"),
    // (Conjunction, "in case that"),
    // (Conjunction, "in case"),
    // (Conjunction, "in order that"),
    // (Conjunction, "in order"),
    // (Conjunction, "now that"),
    // (Conjunction, "only if"),
    // (Conjunction, "provided that"),
    // (Conjunction, "rather than"),
    // (Conjunction, "so that"),

    (Particle, "by"),
    (Particle, "in"),
    (Particle, "not"),
    (Particle, "on"),
    (Particle, "to"),
    (Particle, "oh"),
];

const UTF_COMPOSE_MAP: &[(&'static str, &'static str)] = &[];

const UTF_REDUCE_MAP: &[(&'static str, &'static str)] = &[];


pub fn lang_english() -> Lang {
    let mut lang = Lang::new();

    lang.set_stemmer(Some(Stemmer::create(Algorithm::English)));

    for (from, to) in UTF_COMPOSE_MAP { lang.add_unicode_composition(from, to); }
    for (from, to) in UTF_REDUCE_MAP  { lang.add_unicode_reduction(from, to); }

    for &(pos, word) in FUNCTION_WORDS { lang.add_pos(word, pos); }

    lang
}


#[cfg(test)]
mod tests {
    use crate::utils::to_vec;
    use super::PartOfSpeech;
    use super::{lang_english, UTF_COMPOSE_MAP, UTF_REDUCE_MAP};

    #[test]
    fn stem() {
        let lang = lang_english();
        let w    = to_vec("universe");
        assert_eq!(lang.stem(&w), 7);
    }

    #[test]
    fn get_pos() {
        let lang = lang_english();
        let w1   = to_vec("universe");
        let w2   = to_vec("the");
        assert_eq!(lang.get_pos(&w1), None);
        assert_eq!(lang.get_pos(&w2), Some(PartOfSpeech::Article));
    }

    #[test]
    fn unicode_compose() {
        let lang   = lang_english();
        let source = to_vec("universe");
        let norm   = lang.unicode_compose(&source);
        assert_eq!(norm, None);
    }

    #[test]
    fn unicode_reduce() {
        let lang   = lang_english();
        let source = to_vec("universe");
        let norm   = lang.unicode_reduce(&source);
        assert_eq!(norm, None);
    }

    #[test]
    fn unicode_compose_map_dimenstions() {
        for &(nfd, nfc) in UTF_COMPOSE_MAP {
            assert_eq!(nfd.chars().count(), 2);
            assert_eq!(nfc.chars().count(), 1);
        }
    }

    #[test]
    fn unicode_reduce_map_dimenstions() {
        for &(normal, reduced) in UTF_REDUCE_MAP {
            assert_eq!(normal .chars().count(), 1, "UTF_REDUCE_MAP['{}'] != 1", normal);
            assert_eq!(reduced.chars().count(), 1, "UTF_REDUCE_MAP['{}'].len() != 1", reduced);
        }
    }
}