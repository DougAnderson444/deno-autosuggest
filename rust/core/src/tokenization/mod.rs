mod word;
mod text;

use crate::lang::{Lang, Chars, CharPattern};
pub use word::Word;
pub use text::Text;


pub fn tokenize_query(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
    Text::from_str(source)
        .normalize(lang)
        .fin(false)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}


pub fn tokenize_record(source: &str, lang: &Option<Lang>) -> Text<Vec<char>> {
     Text::from_str(source)
        .normalize(lang)
        .split(&[Chars::Whitespaces, Chars::Control, Chars::Punctuation])
        .strip(&[Chars::NotAlphaNum])
        .lower()
        .mark_pos(lang)
        .stem(lang)
}
