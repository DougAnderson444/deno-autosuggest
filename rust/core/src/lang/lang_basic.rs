#![allow(dead_code)]

use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_basic() -> Lang {
    let mut lang = Lang::new();
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }
    lang
}
