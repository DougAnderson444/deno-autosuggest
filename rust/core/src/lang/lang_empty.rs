#![allow(dead_code)]

use rust_stemmers::{Algorithm, Stemmer};
use super::Lang;
use super::constants::CHAR_CLASSES_LATIN;

pub fn lang_empty() -> Lang {
    let mut lang = Lang::new();

    lang.set_stemmer(Some(Stemmer::create(Algorithm::English)));
    for &(class, ch) in CHAR_CLASSES_LATIN { lang.add_char_class(ch, class); }

    lang
}
