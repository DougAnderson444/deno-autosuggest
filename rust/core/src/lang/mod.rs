mod normalize;
mod lang_english;
mod lang_german;
mod lang_portuguese;
mod lang_russian;
mod lang_spanish;

pub use lang_german::lang_german;
pub use lang_english::lang_english;
pub use lang_spanish::lang_spanish;
pub use lang_portuguese::lang_portuguese;
pub use lang_russian::lang_russian;

use std::cell::RefCell;
use fnv::{FnvHashMap as HashMap};
use rust_stemmers::Stemmer;
use crate::utils::to_vec;
use crate::tokenization::PartOfSpeech;
use normalize::Normalize;


const BUFFER_CAPACITY: usize = 20;


pub struct Lang {
    stemmer:      Option<Stemmer>,
    pos_map:      HashMap<Vec<char>, PartOfSpeech>,
    compose_map:  HashMap<Vec<char>, Vec<char>>,
    reduce_map:   HashMap<Vec<char>, Vec<char>>,
    stem_buffer:  RefCell<String>,
    norm_buffer1: RefCell<Vec<char>>,
    norm_buffer2: RefCell<Vec<char>>,
}


impl Lang {
    pub fn new() -> Self {
        Self {
            stemmer:      None,
            pos_map:      HashMap::default(),
            compose_map:  HashMap::default(),
            reduce_map:   HashMap::default(),
            stem_buffer:  RefCell::new(String::with_capacity(BUFFER_CAPACITY)),
            norm_buffer1: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
            norm_buffer2: RefCell::new(Vec::with_capacity(BUFFER_CAPACITY)),
        }
    }

    pub fn set_stemmer(&mut self, stemmer: Option<Stemmer>) {
        self.stemmer = stemmer;
    }

    pub fn add_pos(&mut self, word: &str, pos: PartOfSpeech) {
        let source   = to_vec(word);
        let composed = self.unicode_compose(&source).unwrap_or(source);
        let reduced  = self.unicode_reduce(&composed);

        self.pos_map.insert(composed, pos);

        if let Some((_, reduced)) = reduced {
            self.pos_map.insert(reduced, pos);
        }
    }

    pub fn add_unicode_composition(&mut self, from: &str, to: &str) {
        self.compose_map.insert(to_vec(from), to_vec(to));
    }

    pub fn add_unicode_reduction(&mut self, from: &str, to: &str) {
        self.reduce_map.insert(to_vec(from), to_vec(to));
    }

    pub fn stem(&self, word: &[char]) -> usize {
        if let Some(stemmer) = &self.stemmer {
            let buffer = &mut *self.stem_buffer.borrow_mut();
            buffer.clear();
            buffer.extend(word.iter());
            let stem = stemmer.stem(&buffer);
            stem.chars().count()
        } else {
            word.len()
        }
    }

    pub fn get_pos(&self, word: &[char]) -> Option<PartOfSpeech> {
        self.pos_map.get(word).cloned()
    }

    pub fn unicode_compose(&self, word: &[char]) -> Option<Vec<char>> {
        let buffer = &mut *self.norm_buffer1.borrow_mut();
        buffer.clear();

        for (_, norm_chunk) in Normalize::new(word, &self.compose_map) {
            buffer.extend(norm_chunk);
        }

        if &buffer[..] == word {
            None
        } else {
            Some(buffer.clone())
        }
    }

    pub fn unicode_reduce(&self, word: &[char]) -> Option<(Vec<char>, Vec<char>)> {
        let buffer1 = &mut *self.norm_buffer1.borrow_mut();
        let buffer2 = &mut *self.norm_buffer2.borrow_mut();
        buffer1.clear();
        buffer2.clear();

        for (word_chunk, norm_chunk) in Normalize::new(word, &self.reduce_map) {
            buffer1.extend(word_chunk);
            buffer2.extend(norm_chunk);
            for _ in 0 .. norm_chunk.len() - word_chunk.len() {
                buffer1.push('\0');
            }
        }

        if &buffer2[..] == word {
            None
        } else {
            Some((buffer1.clone(), buffer2.clone()))
        }
    }
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::utils::to_vec;
    use crate::tokenization::PartOfSpeech;
    use super::Lang;

    fn get_lang() -> Lang {
        let mut lang = Lang::new();
        lang.add_unicode_composition("ó", "ó");
        lang.add_unicode_reduction("ó", "o");
        lang.add_unicode_reduction("õ", "oo");
        lang.add_pos("fóo", PartOfSpeech::Particle);
        lang
    }

    #[test]
    fn unicode_compose_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_compose_nfd() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_compose(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_reduced() {
        let input  = to_vec("foobar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_nfc() {
        let input  = to_vec("foóbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn unicode_reduce_fill0() {
        let input  = to_vec("fõbar");
        let output = get_lang().unicode_reduce(&input[..]);
        assert_debug_snapshot!(output);
    }

    #[test]
    fn get_pos_source() {
        let input = to_vec("fóo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_reduced() {
        let input = to_vec("foo");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, Some(PartOfSpeech::Particle));
    }

    #[test]
    fn get_pos_unknown() {
        let input = to_vec("bar");
        let pos   = get_lang().get_pos(&input);
        assert_eq!(pos, None);
    }
}