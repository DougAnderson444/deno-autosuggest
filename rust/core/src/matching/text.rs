use std::collections::HashSet;
use crate::tokenization::Text;
use super::{WordMatch, word_match};


pub fn text_match(rtext: &Text<&[char]>, qtext: &Text<&[char]>) -> Vec<WordMatch> {
    let capacity = min!(rtext.words.len(), qtext.words.len());
    let mut taken: HashSet<usize> = HashSet::with_capacity(capacity);
    let mut matches: Vec<WordMatch> = Vec::with_capacity(capacity);

    for (i, qword) in qtext.words.iter().enumerate() {
        let mut found: Option<WordMatch> = None;

        for (j, rword) in rtext.words.iter().enumerate() {
            if taken.contains(&j) { continue; }
            if let Some(mut m) = word_match(rword, qword, &rtext.chars, &qtext.chars) {
                m.query.ix  = i;
                m.record.ix = j;
                if found.is_none() && !m.record.primary {
                    found = Some(m);
                    continue;
                }
                if m.record.primary {
                    found = Some(m);
                    break;
                }
            }
        }

        if let Some(m) = found {
            taken.insert(m.record.ix);
            matches.push(m);
        }
    }

    matches
}


#[cfg(test)]
mod tests {
    use insta::assert_debug_snapshot;
    use crate::tokenization::{Chars, Text};
    use crate::lang::lang_english;
    use super::{text_match};


    fn text(s: &str) -> Text<Vec<char>> {
        Text::from_str(s).split(&Chars::Whitespaces)
    }


    #[test]
    fn match_text_empty_both() {
        let qtext = text("").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_empty_one() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
        assert_debug_snapshot!(text_match(&qtext.to_ref(), &rtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_equality() {
        let qtext = text("mailbox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_singleton_typos() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_first() {
        let qtext = text("yelow").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_second() {
        let qtext = text("maiblox").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_pair_unfinished() {
        let qtext = text("maiblox yel").fin(false);
        let rtext = text("yellow mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_intersection() {
        let qtext = text("big malibox yelo").fin(false);
        let rtext = text("small yellow metal mailbox");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_best_rword() {
        let mut qtext = text("the").fin(false);
        let mut rtext = text("the theme");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));

        let lang = lang_english();
        qtext = qtext.mark_pos(&lang);
        rtext = rtext.mark_pos(&lang);
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }


    #[test]
    fn match_text_regression_best_match() {
        let qtext = text("sneak").fin(false);
        let rtext = text("sneaky");
        assert_debug_snapshot!(text_match(&rtext.to_ref(), &qtext.to_ref()));
    }
}
