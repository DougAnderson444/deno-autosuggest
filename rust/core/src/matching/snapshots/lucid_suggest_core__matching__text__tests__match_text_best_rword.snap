---
source: src/matching/text.rs
expression: "text_match(&rtext.to_ref(), &qtext.to_ref())"
---
[
    WordMatch { [rrr] /0/ [qqq] },
]
